### PR TITLE
Introduce ListenerStartHandler in the listener runner for better managing Django DB connections

### PR DESCRIPTION
**Pull request by seratch**
Long story short, this pull request is a revised solution of #512 

This patch can resolve the potential resource issue with Django ORM. The ORM automatically acquires a database connection when performing database queries in a thread but if the thread is not managed by Django Web framework, such connections won't be properly released.

In the pull request https://github.com/slackapi/bolt-python/pull/512, I was wanting to resolve this issue by changing only the internals of the Django adapter. However, thanks to @ross's great feedback https://github.com/slackapi/bolt-python/pull/512#issuecomment-958344238, we found that the approach does not work as expected. The clean up method calls must be done in the thread that is used for asynchronous listener execution.

To resolve this issue, we have to add a new hook point called `listener_start_handler` in the `ThreadListenerRunner`. We already have `listener_error_handler` and `listener_completion_handler`. In a similar way, we'll add a new handler that is supposed to be called at the beginning of `ThreadListenerRunner`. The handler can perform `django.db. close_old_connections()` method in the expected thread, plus before the listener execution.

In the sake of maintaining the feature parity in the asyncio-based `AsyncApp`, this pull request adds `AsyncListenerStartHandler` as well. But the asyncio-based ones are unrelated to the Django issue.

@ross 
Thank you very much for your help (again). If you have any comments on this pull request, your inputs would be greatly appreciated.

### Category (place an `x` in each of the `[ ]`)

* [x] `slack_bolt.App` and/or its core components
* [x] `slack_bolt.async_app.AsyncApp` and/or its core components
* [x] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [ ] Others

## Requirements (place an `x` in each `[ ]`)

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: slack_bolt/listener/async_listener_completion_handler.py [hunk 1]
```diff
@@ -15,10 +15,9 @@ async def handle(
         request: AsyncBoltRequest,
         response: Optional[BoltResponse],
     ) -> None:
-        """Handles an unhandled exception.
```
Comment (seratch) [inline slack_bolt/listener/async_listener_completion_handler.py:18]: This had been wrong 🤦

File: slack_bolt/listener/listener_completion_handler.py [hunk 2]
```diff
@@ -15,7 +15,7 @@ def handle(
         request: BoltRequest,
         response: Optional[BoltResponse],
     ) -> None:
-        """Handles an unhandled exception.
+        """Do something extra after the listener execution
```
Comment (seratch) [inline slack_bolt/listener/listener_completion_handler.py:18]: This had been wrong 🤦

File: slack_bolt/version.py [hunk 3]
```diff
@@ -1,2 +1,2 @@
 """Check the latest version at https://pypi.org/project/slack-bolt/"""
-__version__ = "1.9.4"
+__version__ = "1.10.0a"
```
Comment (seratch) [inline slack_bolt/version.py:2]: We'll bump the minor version as this pull request introduces new functionalities

**Review by filmaj — COMMENTED** [review]
... given I approved the last PR that led to the thrice-execution issue described by @ross, not sure my review should count for much 🙈 😓 

I need to first familiarize myself with Django + Bolt 🤐

**Comment by ross** [comment]
I'm not seeing two calls fro the right thread with the expected `timing`s. Still doing a bit of testing to make sure they're at the expected points.

```
2021-11-03T15:59:50Z DEBUG Released thread-bound old DB connections (thread name: bot-worker_0, execution timing: listener-start)
...
2021-11-03T15:59:51Z DEBUG Released thread-bound old DB connections (thread name: bot-worker_0, execution timing: listener-completion)
```

The `ThreadPoolExecutor` being created when one is not provided is using the default names on the worker threads. https://github.com/slackapi/bolt-python/blob/cf3951e4eebe7f72daf235d285023fbda090d24a/slack_bolt/app/app.py#L311

Adding `thread_name_prefix=slack_bolt-workers` or similar would give those threads better names to more easily identify them when digging into things like this. I'm passing my own `ThreadPoolExecutor` in above which is why mine have names.

**Comment by ross** [comment]
Verified that the calls are wrapping the `app.event` call. So I think we're good there. I also went ahead and verified the calls happening to `close_old_connections` including threads and I think they're as expected:

```
2021-11-03T16:13:45Z INFO  db close_old_connections: thread=Thread-1
...
2021-11-03T16:13:46Z INFO  db close_old_connections: thread=simone-worker_0
...
2021-11-03T16:13:46Z INFO  db close_old_connections: thread=Thread-1
...
2021-11-03T16:13:48Z INFO  db close_old_connections: thread=simone-worker_0
```

Based on stack traces the `Thread-1` connections are coming from django's `runserver` and are the http threads not owned by `slack_bolt` so noise w/respect to this PR.

There's a 3rd pair that happens just after that. It appears to be coming from the self message triggered by my bot's reply so that might have been the source of the `Thread-2` second one I couldn't account for in https://github.com/slackapi/bolt-python/pull/512#issuecomment-958610514.

I can't do a formal approval or anything so 👍 will have to count :grin:

**Comment by seratch** [comment]
@ross @filmaj Thanks a lot for quickly checking this PR!

**Comment by seratch** [comment]
I will make a new minor version release later today

**Pull request merged**